trio/tests.py

```diff
@@ -0,0 +1,31 @@
+import importlib
+import sys
+from typing import Any
+
+from . import _tests
+from ._deprecate import warn_deprecated
+
+
+# This won't give deprecation warning on import, but will give a warning on use of any
+# attribute in tests, and static analysis tools will also not see any content inside.
+class TestsDeprecationWrapper:
+    __name__ = "trio.tests"
+
+    def __getattr__(self, attr: str) -> Any:
+        warn_deprecated(
+            f"trio.tests.{attr}",
+            "0.24.0",
+            instead=f"trio._tests.{attr}",
+            issue="https://github.com/python-trio/trio/issues/274",
+        )
+
+        # needed to access e.g. trio._tests.tools, although pytest doesn't need it
+        if not hasattr(_tests, attr):  # pragma: no cover
+            importlib.import_module(f"trio._tests.{attr}", "trio._tests")
+            return attr
+
+        return getattr(_tests, attr)
+
+
+# https://stackoverflow.com/questions/2447353/getattr-on-a-module
+sys.modules[__name__] = TestsDeprecationWrapper()  # type: ignore[assignment]
```

Comment:
I'm not sure we really need to be this careful about backcompat here -- *surely* people aren't *actually* using `trio.tests`, right? -- but I guess it's harmless.

Though... I guess we actually want to get rid of `trio.tests` entirely, so maybe we should issue the warning on import? If someone is importing it but not using it (for some reason), then they won't get a warning, but they'll still get broken in the future when we delete this file.

Reply:
ah, took me a minute but figured out how to get it to warn on import as well. Only thing it doesn't warn on now that would break is `import trio; assert trio.tests` - i.e. access of the module, but without importing it, and not accessing any of it's attributes. (I tried changing from using `__getattr__` to `__getattribute__` but that made no difference in this case). There's probably some way of addressing that, but now we're getting into very weird cases.